File: neetcode all and misc questions/Linked Lists/copy_list_with_random_pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        """
        https://neetcode.io/problems/copy-linked-list-with-random-pointer
        Revisited 9/30/2024
        This solution appears to be cleaner than my last solution, using only 2 maps instead of 3

        2 Hashmaps:
        hashmap #1: the key is the old pointer memory address, and the value
        is its new LL node.
        hashmap #2: the new LL memory address is mapped to old Node
        create a new linked list with the next nodes, at the same time,
        we build hashmap #2, mapping id(newNode) to old node and hashmap #1, mapping id(old node) to newNode
        Loop through the new linked list again, but this time, call id() on the new linked list node,
        and use the id access hashmap #2 to find the old counterpart. Then, use the id() of the old counterpart's random pointer
        to access hashmap #1, which will help determine which new LL node the random pointer
        should point to. 
        Note that if the old counterpart's random pointer is pointing to None, just set the new node's random
        to be None as well so we don't call id() on None
        """
        temp1 = head
        newList = Node(0)
        map1 = {}
        map2 = {}
        ptr = newList
        while (temp1 != None):
            newNode = Node(temp1.val)
            ptr.next = newNode
            map2[id(newNode)] = temp1
            map1[id(temp1)] = newNode
            ptr = ptr.next
            temp1 = temp1.next
        temp2 = newList.next
        while (temp2 != None):
            old = map2[id(temp2)]
            if (old.random):
                newValForRandom = map1[id(old.random)]
                temp2.random = newValForRandom
            else:
                temp2.random = None
            temp2 = temp2.next
        return newList.next

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        """
        the node's values don't have to be unique, so we can't
        rely on the node's values, we can use the memory addresses instead (id function)
         
        Key Concepts:
        1) We need three hashmaps:
        
        oldAddresses = mapping old node id to it's old random id (if old node doesn't have a random, map it to null)
        newAddresses = map the newNode id to the oldNode id
        
        idToNode = map the oldNode id to the actual newNode itself 
        
        2) Loop through the old linked list, creating the new linked list using the dummy node strategy, and mapping all the appropriate
        values into their hashmaps
        
        3) Using the newly created linked list, we take the id of each node, 
           take the new node id and find the old node id using newAddresses
           
           Then, take the old node id and find it's old random node id using oldAddresses
           
           Then, take the old random node id and find it's new node equivalent using idToNode
           
           Then, map our new node's random to the equivalent found in idToNode
           
           If the old random node id was None, just map new node's random to None
         
         4) Return the new head
           
         Time complexity: O(N)
         Space Complexity: O(N)
        
        
        """
        oldAddresses = dict()
        newAddresses = dict()
        idToNode = dict()
        oldHead = head
        dummy = ListNode(0)
        newHead = dummy
        # map the old node to it's old random memory address
        # map the new node id to the old node id
        # map the old node id to it's new node equivalent 
        while (oldHead != None):
            oldNodeAddress = id(oldHead)
            if (oldHead.random):
                oldRandomAddress = id(oldHead.random)
            else:
                oldRandomAddress = None
            oldAddresses[oldNodeAddress] = oldRandomAddress
            newNode = ListNode(oldHead.val)
            newHead.next = newNode
            newAddresses[id(newNode)] = oldNodeAddress
            idToNode[oldNodeAddress] = newNode
            oldHead = oldHead.next
            newHead = newHead.next
        # go back and map the new random nodes on the newly copied linked list
        resHead = dummy.next
        cur = resHead
        while (cur):
            idOldNode = newAddresses[id(cur)]
            oldRandomAddress = oldAddresses[idOldNode]
            if oldRandomAddress != None:
                newNode = idToNode[oldRandomAddress]
            else:
                newNode = None
            cur.random = newNode
            cur = cur.next
        return resHead
        

File: node_js_ds_and_algos_implementations/linked lists/copy_list_with_random_pointer.js
// class Node {
//   constructor(val, next = null, random = null) {
//       this.val = val;
//       this.next = next;
//       this.random = random;
//   }
// }

class Solution {
    /**
     * @param {Node} head
     * @return {Node}
     */
    copyRandomList(head) {
        let temp = head
        /* 
            4/23/2026
            This is a much cleaner solution than the original that I came up with:

            map the old node reference to its new node reference 
            after we've created all new nodes to match their old nodes,
            we iterate through the old nodes once more
                and then determine what the next and random pointers to the new nodes
                are based on the references of the old nodes
            
            Important NOTE: use new Map() instead of a regular object, since regular objects
            will coerce the node references into strings instead of keeping them as their 
            actual references in memory. This would cause the object to constantly overwrite
            the same key over and over since they all coerce to "[Object object]" instead
            of the actual reference
        */
        let oldToNew = new Map()
        oldToNew.set(null,null)
        while (temp){
            let newNode = new Node(temp.val, null, null)
            oldToNew.set(temp, newNode)
            temp = temp.next
        }
        temp = head
        let res = new Node()
        let dummy = res
        while (temp){
            let copy = oldToNew.get(temp)
            // after retrieving the new node based on the old,
            // set the next pointer for the new node by matching the old pointers' next 
            copy.next = oldToNew.get(temp.next)
            // set the random pointer for the new node by matching the old pointers' random 
            copy.random = oldToNew.get(temp.random)
            dummy.next = copy
            dummy = dummy.next
            temp = temp.next
        }
        return res.next
    }
}

/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {
   /*
   2 references:
   array of each list node, so the node can be retrieved through its index
   hashmap: key is the index of the ith node, and the value is the index of the node pointed by the random pointer
   
   7 -> 13 -> 11 -> 10 -> 1 -> null
   
   1) push each of the original list nodes into the array so we have the index
   2) Loop through the list again, but this time, use findIndex to find the node that's being 
   referenced by random pointer
   3) Add the entry to the hashmap, where the key is the current index, and the value is the index
   of the node that's being pointed to by random pointer
   
   for example,
   if 13 points to 7, this would be the equivalent of {1: 0}, because i = 1 node points to i = 0 node
   
   4) At the same time as step 3, construct the copied list using a dummy pointer
   5) Repeat step 1 on the new copied list
   6) Once we have the newNodeList, we loop through newNodeList, and then reference back to our hashmap
   
   if our current node index has a value of -1 in the hashmap, that means it originally pointed to null (as -1 means the index could not be found)
   otherwise, it should point to the index somewhere in newNodeList, so we set random pointer to be newNodeList[randPointerMap[i]], which should
   be the proper location that it was pointing to originally.
   */ 
    let nodeList = []
    let temp = head
    while (temp){
        nodeList.push(temp)
        temp = temp.next
    }
    let temp2 = head
    let randPointerMap = {}
    // at this time, also create the copy of the linked list through the dummy node
    let newNodeList = []
    let dummy = new ListNode(0)
    let newHead = dummy
    for (let i = 0; i < nodeList.length; ++i){
        newNode = new ListNode(nodeList[i].val)
        newHead.next = newNode
        newHead = newHead.next
        let randIndex = nodeList.findIndex(x => x === nodeList[i].random)
        randPointerMap[i] = randIndex
    }

    let temp3 = dummy.next
    // append each node to the new node list so we can retrieve the node by index
    while (temp3){
        newNodeList.push(temp3)
        temp3 = temp3.next
    }
    // loop through the new node list, this time setting the random pointers
    for (let i = 0; i < newNodeList.length; ++i){
        let index = randPointerMap[i]
        if (index === -1){
            newNodeList[i].random = null
        }
        else {
            newNodeList[i].random = newNodeList[index]
        }
    }
    return dummy.next
};